src/cryptsetup/cryptsetup-tokens/luks2-fido2.c

```diff
@@ -0,0 +1,150 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include <libcryptsetup.h>
+
+#include "cryptsetup-token-util.h"
+#include "hexdecoct.h"
+#include "json.h"
+#include "luks2-fido2.h"
+#include "memory-util.h"
+
+int acquire_luks2_key(
+                struct crypt_device *cd,
+                const char *json,
+                const char *device,
+                char **pins,
+                char **ret_keyslot_passphrase,
+                size_t *ret_keyslot_passphrase_size) {
+
+        int r;
+        Fido2EnrollFlags required;
+        size_t cid_size, salt_size, decrypted_key_size;
+        _cleanup_(json_variant_unrefp) JsonVariant *v = NULL;
+        _cleanup_free_ void *cid = NULL, *salt = NULL;
+        _cleanup_free_ char *rp_id = NULL;
+        _cleanup_(erase_and_freep) void *decrypted_key = NULL;
+        _cleanup_(erase_and_freep) char *base64_encoded = NULL;
+
+        assert(ret_keyslot_passphrase);
+        assert(ret_keyslot_passphrase_size);
+
+        r = parse_luks2_fido2_data(cd, json, &rp_id, &salt, &salt_size, &cid, &cid_size, &required);
+        if (r < 0)
+                return r;
+
+        /* configured to use pin but none was provided */
+        if ((required & FIDO2ENROLL_PIN) && !pins)
```

Comment:
!pins → strv_isempty(pins)

Reply:
Just out of curiosity: Empty PIN string ("") is not valid PIN in both fido2 and pkcs11?

Reply:
so far we didn't allow that. I'd wait until someone complains about this. i'd rather be safe than sorry.

for both fido2 and pkcs11 things are typically strings, not binary blobs, hence I think not accepting an empty string is OKish.

Reply:
Thanks, makes sense.